nilearn/glm/first_level/first_level.py

```diff
@@ -1060,6 +1034,249 @@ def first_level_from_bids(dataset_path, task_label, space_label=None,
+                             space_label: str | None,
+                             img_filters: list[Tuple[str, str]],
+                             derivatives_path: str,
+                             supported_filters: list[str],
+                             ) -> None:
+    """Check type and value of arguments of first_level_from_bids.
+
+    Parameters
+    ----------
+    dataset_path : :obj:`str`
+        _description_
+    task_label : :obj:`str`
+        _description_
+    space_label : :obj:`str` or None
+        _description_
+    img_filters : list[Tuple[str, str]]
+        _description_
+    derivatives_path : :obj:`str`
+        _description_
+    supported_filters : list[str]
+        _description_
+
+    Raises
+    ------
+    TypeError
+        _description_
+    ValueError
+        _description_
+    """    
+    if not isinstance(dataset_path, str):
+        raise TypeError(
+            "dataset_path must be a string, "
+            f"instead {type(dataset_path)} was given"
+        )
+    if not os.path.exists(dataset_path):
+        raise ValueError(f'given path do not exist: {dataset_path}')
+
+    if not os.path.exists(derivatives_path):
+        raise ValueError('derivatives folder does not exist in given dataset')
+
+    if not isinstance(task_label, str):
+        raise TypeError(
+            "task_label must be a string, "
+            f"instead {type(task_label)} was given"
+        )
+
+    if space_label is not None and not isinstance(space_label, str):
+        raise TypeError(
+            "space_label must be a string, "
+            f"instead {type(space_label)} was given."
+        )
+
+    if not isinstance(img_filters, list):
+        raise TypeError(
+            f"img_filters must be a list, "
+            f"instead {type(img_filters)} was given."
+        )
+    for this_filter in img_filters:
+        if (not isinstance(this_filter[0], str)
+                or not isinstance(this_filter[1], str)):
+            raise TypeError('filters in img filters must be (str, str), '
+                            f"instead {type(this_filter)} was given.")
+        if this_filter[0] not in supported_filters:
+            raise ValueError(
+                f"field {this_filter[0]} is not a possible filter. "
+                f"Only {supported_filters} are allowed.")
+
+
+def _bids_filter(task_label: str,
+                 space_label: str | None = None,
+                 supported_filters: list[str] | None = None,
+                 extra_filter: list[Tuple[str, str]] | None = None
+                 ) -> list[Tuple[str, str]]:
+    """Return a filter to specific files from a BIDS dataset.
+
+    Parameters
+    ----------
+    task_label : :obj:`str`
+        _description_
+    space_label : :obj:`str` or None, optional
+        _description_
+    supported_filters : list[str] or None, optional
+        _description_
+    extra_filter : list[Tuple[str, str]] or None, optional
```

Comment:
I was trying to use same type as `img_filters` that is part of the pubic API and is also a list of tuple of (str, str) (not sure why).

But I agree that a dict makes more sense. 
Most likely will implement everything with dict.
I would even be tempted to allow a dict in the public API and deprecate the list of tuple in the future.

Reply:
Actually realizing that the reason I did this is because the docstring for `get_bids_file` (that the filters are passed to) reads:

```python
   """
   ...

    filters : :obj:`list` of :obj:`tuple` (:obj:`str`, :obj:`str`), optional
        Filters are of the form (field, label). Only one filter per field
        allowed. A file that does not match a filter will be discarded.
        Filter examples would be ('ses', '01'), ('dir', 'ap') and
        ('task', 'localizer').
   """
```

but the implementation seems compatible with dict or tuple (via tuple unpacking).

```python
    if filters:
        files = [parse_bids_filename(file_) for file_ in files]
        for key, value in filters:
            files = [
                file_ for file_ in files if
                (key in file_ and file_[key] == value)
            ]
        return [ref_file['file_path'] for ref_file in files]
```

If I read this correctly the doc string should probably be updated.

Reply:
Actually not, this does require a list of tuples to work.

So unless we start changing the expected type of several public functions I am going to keep things as is rather than change from list of tuples to a dict and then back to a list of tuples.

OK to implement those changes but this should be done in a separate PR as it may involve some breaking changes.